File: src/wrap-openai.ts
import type { ChatCompletion, ChatCompletionChunk } from 'openai/resources/chat/completions';
import { WatchOptions, LogPayload } from './types';
import { calcCost } from './cost';
import { sendLog } from './ingest';

function extractPrompt(params: Record<string, unknown>): string {
  return ((params.messages as any[]) ?? [])
    .map((m: any) => {
      const content =
        typeof m.content === 'string'
          ? m.content
          : Array.isArray(m.content)
          ? m.content.map((c: any) => c.text ?? '').join('')
          : '';
      return `[${m.role}]: ${content}`;
    })
    .join('\n');
}

function errorPayload(
  start: number,
  model: string,
  prompt: string,
  err: unknown,
): LogPayload {
  return {
    timestamp:     new Date(start).toISOString(),
    model,
    prompt,
    completion:    '',
    latency_ms:    Date.now() - start,
    tokens_used:   0,
    cost_usd:      0,
    error_type:    err instanceof Error ? err.constructor.name : 'UnknownError',
    error_message: err instanceof Error ? err.message : String(err),
  };
}

async function* interceptStream(
  stream: AsyncIterable<ChatCompletionChunk>,
  start: number,
  params: Record<string, unknown>,
  options: WatchOptions,
): AsyncGenerator<ChatCompletionChunk> {
  let completion = '';
  let inputTokens = 0;
  let outputTokens = 0;
  let model = params.model as string;
  let thrownError: unknown;

  try {
    for await (const chunk of stream) {
      yield chunk;
      if (chunk.model) model = chunk.model;
      const delta = chunk.choices?.[0]?.delta?.content;
      if (delta) completion += delta;
      // stream_options: { include_usage: true } makes usage appear on the final chunk
      if (chunk.usage) {
        inputTokens  = chunk.usage.prompt_tokens     ?? 0;
        outputTokens = chunk.usage.completion_tokens ?? 0;
      }
    }
  } catch (err) {
    thrownError = err;
    throw err;
  } finally {
    if (thrownError !== undefined) {
      sendLog(errorPayload(start, model, extractPrompt(params), thrownError), options);
    } else {
      sendLog({
        timestamp:   new Date(start).toISOString(),
        model,
        prompt:      extractPrompt(params),
        completion,
        latency_ms:  Date.now() - start,
        tokens_used: inputTokens + outputTokens,
        cost_usd:    calcCost(model, inputTokens, outputTokens),
      }, options);
    }
  }
}

/**
 * Patches client.chat.completions.create in-place to intercept every call.
 * Returns the same client so callers can chain: const ai = wrapOpenAI(new OpenAI(), opts)
 */
export function wrapOpenAI<T extends { chat: { completions: { create: (...args: any[]) => any } } }>(
  client: T,
  options: WatchOptions,
): T {
  const orig = client.chat.completions.create.bind(client.chat.completions);

  (client.chat.completions as any).create = async function (
    params: Record<string, unknown>,
    requestOptions?: unknown,
  ): Promise<unknown> {
    const start = Date.now();

    if (params.stream) {
      try {
        const stream = await orig(params, requestOptions);
        return interceptStream(stream, start, params, options);
      } catch (err) {
        sendLog(errorPayload(start, params.model as string, extractPrompt(params), err), options);
        throw err;
      }
    }

    try {
      const response: ChatCompletion = await orig(params, requestOptions);
      const inputTokens  = response.usage?.prompt_tokens     ?? 0;
      const outputTokens = response.usage?.completion_tokens ?? 0;
      const model        = response.model ?? (params.model as string);

      sendLog({
        timestamp:   new Date(start).toISOString(),
        model,
        prompt:      extractPrompt(params),
        completion:  response.choices?.[0]?.message?.content ?? '',
        latency_ms:  Date.now() - start,
        tokens_used: inputTokens + outputTokens,
        cost_usd:    calcCost(model, inputTokens, outputTokens),
      }, options);

      return response;
    } catch (err) {
      sendLog(errorPayload(start, params.model as string, extractPrompt(params), err), options);
      throw err;
    }
  };

  return client;
}

File: python-sdk/loglens/_wrap_openai.py
import datetime
import time
from typing import Any, Iterator

from ._cost import calc_cost
from ._ingest import send_log
from ._types import LogPayload, WatchOptions


def _extract_prompt(kwargs: dict) -> str:
    parts = []
    for msg in kwargs.get('messages', []):
        content = msg.get('content', '')
        if isinstance(content, str):
            parts.append(content)
        elif isinstance(content, list):
            for block in content:
                if isinstance(block, dict) and block.get('type') == 'text':
                    parts.append(block.get('text', ''))
    return ' '.join(parts)[:2000]


def _now() -> str:
    return datetime.datetime.utcnow().isoformat() + 'Z'


def _error_payload(start: float, model: str, prompt: str, exc: Exception) -> LogPayload:
    return LogPayload(
        timestamp=_now(),
        model=model,
        prompt=prompt,
        completion='',
        latency_ms=int((time.time() - start) * 1000),
        tokens_used=0,
        cost_usd=0.0,
        error_type=type(exc).__name__,
        error_message=str(exc),
    )


class _OpenAIStreamWrapper:
    """Wraps a Stream[ChatCompletionChunk] to collect and log after exhaustion."""

    def __init__(self, stream: Any, start: float, model: str, prompt: str, opts: WatchOptions):
        self._stream = stream
        self._start = start
        self._model = model
        self._prompt = prompt
        self._opts = opts

    def __iter__(self) -> Iterator:
        completion_parts = []
        input_tokens = 0
        output_tokens = 0
        thrown = None

        try:
            for chunk in self._stream:
                choices = getattr(chunk, 'choices', [])
                if choices:
                    delta = getattr(choices[0], 'delta', None)
                    content = getattr(delta, 'content', None) if delta else None
                    if content:
                        completion_parts.append(content)

                # usage is sent in the final chunk when stream_options.include_usage=True
                usage = getattr(chunk, 'usage', None)
                if usage:
                    input_tokens = getattr(usage, 'prompt_tokens', 0)
                    output_tokens = getattr(usage, 'completion_tokens', 0)

                yield chunk
        except Exception as exc:
            thrown = exc
            raise
        finally:
            latency = int((time.time() - self._start) * 1000)
            if thrown:
                payload = _error_payload(self._start, self._model, self._prompt, thrown)
                payload.latency_ms = latency
            else:
                total = input_tokens + output_tokens
                payload = LogPayload(
                    timestamp=_now(),
                    model=self._model,
                    prompt=self._prompt,
                    completion=''.join(completion_parts),
                    latency_ms=latency,
                    tokens_used=total,
                    cost_usd=calc_cost(self._model, input_tokens, output_tokens),
                )
            send_log(payload, self._opts)

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if hasattr(self._stream, '__exit__'):
            return self._stream.__exit__(*args)
        return False

    def __getattr__(self, name: str):
        return getattr(self._stream, name)


class _WrappedCompletions:
    def __init__(self, completions: Any, opts: WatchOptions):
        self._completions = completions
        self._opts = opts

    def create(self, **kwargs):
        start = time.time()
        model: str = kwargs.get('model', 'unknown')
        prompt = _extract_prompt(kwargs)

        if kwargs.get('stream'):
            raw = self._completions.create(**kwargs)
            return _OpenAIStreamWrapper(raw, start, model, prompt, self._opts)

        try:
            resp = self._completions.create(**kwargs)
            latency = int((time.time() - start) * 1000)

            choices = getattr(resp, 'choices', [])
            completion = ''
            if choices:
                msg = getattr(choices[0], 'message', None)
                completion = getattr(msg, 'content', '') or ''

            usage = getattr(resp, 'usage', None)
            input_tokens = getattr(usage, 'prompt_tokens', 0) if usage else 0
            output_tokens = getattr(usage, 'completion_tokens', 0) if usage else 0

            payload = LogPayload(
                timestamp=_now(),
                model=model,
                prompt=prompt,
                completion=completion,
                latency_ms=latency,
                tokens_used=input_tokens + output_tokens,
                cost_usd=calc_cost(model, input_tokens, output_tokens),
            )
            send_log(payload, self._opts)
            return resp
        except Exception as exc:
            send_log(_error_payload(start, model, prompt, exc), self._opts)
            raise

    def __getattr__(self, name: str):
        return getattr(self._completions, name)


class _WrappedChat:
    def __init__(self, chat: Any, opts: WatchOptions):
        self._chat = chat
        self.completions = _WrappedCompletions(chat.completions, opts)

    def __getattr__(self, name: str):
        return getattr(self._chat, name)


class WrappedOpenAI:
    def __init__(self, client: Any, opts: WatchOptions):
        self._client = client
        self.chat = _WrappedChat(client.chat, opts)

    def __getattr__(self, name: str):
        return getattr(self._client, name)


def wrap_openai(client: Any, options: WatchOptions) -> WrappedOpenAI:
    return WrappedOpenAI(client, options)
